third_party/accessibility/ax/platform/ax_platform_node_textrangeprovider_win.cc

```diff
@@ -0,0 +1,1701 @@
+    BASE_DCHECK(SUCCEEDED(hr));
+    mixed_attribute_value_variant.Set(mixed_attribute_value.Get());
+  }
+
+  base::win::ScopedVariant not_supported_value_variant;
+  {
+    Microsoft::WRL::ComPtr<IUnknown> not_supported_value;
+    HRESULT hr = ::UiaGetReservedNotSupportedValue(&not_supported_value);
+    BASE_DCHECK(SUCCEEDED(hr));
+    not_supported_value_variant.Set(not_supported_value.Get());
+  }
+
+  return !vector.Compare(mixed_attribute_value_variant) ||
+         !vector.Compare(not_supported_value_variant);
+}
+
+// static
+bool AXPlatformNodeTextRangeProviderWin::ShouldReleaseTextAttributeAsSafearray(
+    TEXTATTRIBUTEID attribute_id,
+    const base::win::VariantVector& attribute_value) {
+  // |vector| may be pre-populated with a UIA reserved value. In such a case, we
+  // must release as a scalar variant.
+  return TextAttributeIsArrayType(attribute_id) &&
+         !TextAttributeIsUiaReservedValue(attribute_value);
+}
+
+AXPlatformNodeTextRangeProviderWin::TextRangeEndpoints::TextRangeEndpoints() {
+  start_ = AXNodePosition::CreateNullPosition();
+  end_ = AXNodePosition::CreateNullPosition();
+}
+
+AXPlatformNodeTextRangeProviderWin::TextRangeEndpoints::~TextRangeEndpoints() {
+  SetStart(AXNodePosition::CreateNullPosition());
+  SetEnd(AXNodePosition::CreateNullPosition());
+}
+
+void AXPlatformNodeTextRangeProviderWin::TextRangeEndpoints::SetStart(
+    AXPositionInstance new_start) {
+  bool did_tree_change = start_->tree_id() != new_start->tree_id();
+  // TODO(bebeaudr): We can't use IsNullPosition() here because of
+  // https://crbug.com/1152939. Once this is fixed, we can go back to
+  // IsNullPosition().
+  if (did_tree_change && start_->kind() != AXPositionKind::NULL_POSITION &&
+      start_->tree_id() != end_->tree_id()) {
+    RemoveObserver(start_);
+  }
+
+  start_ = std::move(new_start);
+
+  if (did_tree_change && !start_->IsNullPosition() &&
+      start_->tree_id() != end_->tree_id()) {
+    AddObserver(start_);
+  }
+}
+
+void AXPlatformNodeTextRangeProviderWin::TextRangeEndpoints::SetEnd(
+    AXPositionInstance new_end) {
+  bool did_tree_change = end_->tree_id() != new_end->tree_id();
+  // TODO(bebeaudr): We can't use IsNullPosition() here because of
+  // https://crbug.com/1152939. Once this is fixed, we can go back to
+  // IsNullPosition().
+  if (did_tree_change && end_->kind() != AXPositionKind::NULL_POSITION &&
+      end_->tree_id() != start_->tree_id()) {
+    RemoveObserver(end_);
+  }
+
+  end_ = std::move(new_end);
+
+  if (did_tree_change && !end_->IsNullPosition() &&
+      start_->tree_id() != end_->tree_id()) {
+    AddObserver(end_);
+  }
+}
+
+void AXPlatformNodeTextRangeProviderWin::TextRangeEndpoints::AddObserver(
+    const AXPositionInstance& position) {
+  auto tree = AXTreeManagerMap::GetInstance()
+                  .GetManager(position->GetAnchor()->tree()->GetAXTreeID())
+                  ->GetTree();
```

Comment:
It looks like Chromium finds the tree ID using the equivalent of `position->tree_id()`. Can you not use that here? Why not? Otherwise, would it make sense to make `AddObserver`'s and `RemoveObserver`'s input a tree ID instead of a position?

Reply:
It appears the discrepancy is because Chromium recently changed and reverted these methods and their signatures. At the time I imported this code, Chromium used the version of the method seen here which takes the position instance as a parameter, but currently has changed it to take a tree ID instead.